### PR TITLE
EOS-20929: ConfStore Merge: Merge functionality for Conf Store

### PR DESCRIPTION
**Pull request by vastradparayya**
Signed-off-by: Parayya Vastrad <parayya.vastrad@seagate.com>

# Problem Statement
- ConfStore Merge: Merge functionality for Conf Store

# Design
-  [EOS-20929](https://jts.seagate.com/browse/EOS-20929)

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing 
- [x] Confirm that Test Cases are added (for both the cases, fix and feature) [Y/N]: Y
- [x] Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: Y
- [x] Confirm Testing was performed with installed RPM [Y/N]:  Y

# Review Checklist 
  Before posting the PR please ensure
- [x] PR is self reviewed
- [x] Jira is updated
- [ ] Check if the description is clear and explained. 
- [ ] Check Acceptance Criterion is defined. 
- [ ] All the tests performed should be mentioned before Resolving a JIRA. 
- [ ] Verification needs to be done before marked as Closed/Verified 

# Documentation
- [ ] Changes done to WIKI / Confluence page

### CODE REVIEW

File: py-utils/src/utils/conf_store/conf_store.py [hunk 1]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
         for key in key_list:
             self._cache[dst_index].set(key, self._cache[src_index].get(key))
 
+    def merge(self, src_index: str, dest_index: str):
+        """
+        Merges the content of src_index and dest_index file
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:181]: Please highlight where does the result go, i.e. 
Merges the content of src_index and dest_index file into src_index.

File: py-utils/src/utils/conf_store/conf_store.py [hunk 2]
```diff
@@ -225,6 +284,11 @@ def copy(src_index: str, dst_index: str, key_list: list = None):
         Conf._conf.copy(src_index, dst_index, key_list)
         Conf._conf.save(dst_index)
 
+    @staticmethod
+    def merge(src_index: str, dest_index: str):
+        Conf._conf.merge(src_index, dest_index)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:289]: We should add CLI extension as well in separate patch.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 3]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
         for key in key_list:
             self._cache[dst_index].set(key, self._cache[src_index].get(key))
 
+    def merge(self, src_index: str, dest_index: str):
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:179]: Accept optional parameter as keys[]. Only the keys will be merged not everything.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 4]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        if src_index not in self._cache.keys():
+            raise ConfError(errno.EINVAL, "config index %s is not loaded",
+                src_index)
+        if dest_index not in self._cache.keys():
+            raise ConfError(errno.EINVAL, "config index %s is not loaded",
+                dest_index)
+
+        import re
+        # getting max index of list key for adding in dest_index
+        dest_key_max_indexes = [int(i[0]) for i in [re.findall('\d+',key)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:197]: Please ensure there is a space after every comma i the code.

File: py-utils/src/utils/conf_store/conf_store.py [hunk 5]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+            for key in self._cache[dest_index].get_keys()] if i]
+        dest_key_max_indexes = dest_key_max_indexes if dest_key_max_indexes \
+            else [0]
+        max_index = max(dest_key_max_indexes)
+        count = 0
+        for key in self._cache[src_index].get_keys():
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:203]: This is too complex algo. There is an easy way using set and get
```
for key in src.get_keys(): 
    if not key in dst.get_keys(): 
        dst.set(key, src.get(key))
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 6]
```diff
@@ -225,6 +284,11 @@ def copy(src_index: str, dst_index: str, key_list: list = None):
         Conf._conf.copy(src_index, dst_index, key_list)
         Conf._conf.save(dst_index)
 
+    @staticmethod
+    def merge(src_index: str, dest_index: str):
+        Conf._conf.merge(src_index, dest_index)
+        Conf._conf.save(dest_index)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:290]: save should be left to the user. Merge is in memory only. Should be added to the spec.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 7]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+            else [0]
+        max_index = max(dest_key_max_indexes)
+        count = 0
+        for key in self._cache[src_index].get_keys():
+            # add key in dest_index
+            if key not in self._cache[dest_index].get_keys():
+                self._cache[dest_index].set(key,\
+                    self._cache[src_index].get(key))
+            else:
+                values = self.get_values([], self._cache[
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:209]: We dont have to merge values. Its only about the keys.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 7 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:209]: @ujjwalpl as we discussed in call, you said that result should be union of both contents. You gave a example - 
'If conf has 6 entries and .sample has 8 then conf file should have those additional entries. Existing entries should remain.'

File: py-utils/src/utils/conf_store/conf_store.py [hunk 8]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+                            + count) + re.split('\d+', key)[1]
+                        # set key from src_index to dest_index
+                        self._cache[dest_index].set(new_key,
+                            self._cache[src_index].get(key))
+                    else:
+                        # for single key, overwriting dest_index key
+                        self._cache[dest_index].set(key,
+                            self._cache[src_index].get(key))
+
+    def get_values(self, values: list, data: str) -> list:
```
Comment (suryakumar1024) [inline py-utils/src/utils/conf_store/conf_store.py:228]: get_values name gives different meaning here the API purpose is to just get value of particular List. Kindly update  the name of API.

File: py-utils/src/utils/conf_store/conf_store.py [hunk 9]
```diff
@@ -176,6 +176,65 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        for key in data:
```
Comment (suryakumar1024) [inline py-utils/src/utils/conf_store/conf_store.py:229]: @ujjwalpl kindly confirm this.
Do we have to concern about what is inside list while merge operation

Eg. 
config_1 = { 'a' : [1,2,4,5] }
config_2 = { 'a' : [1,5,3,2] }

current implementation returns 

output:
{ 'a' : [1,2,4,5,3] }

Is this correct?

File: py-utils/src/utils/conf_store/conf_store.py (hunk 7 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:209]: @ujjwalpl Should we consider list keys?
Example keys - 
cortx>software>common>message_bus_type
cortx>software>kafka>servers[0]
cortx>software>kafka>servers[1]
cortx>software>kafka>servers[2]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 1 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:181]: Done

File: py-utils/src/utils/conf_store/conf_store.py (hunk 2 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:289]: Done. Created a separate jira ticket

File: py-utils/src/utils/conf_store/conf_store.py (hunk 3 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:179]: Done.

File: py-utils/src/utils/conf_store/conf_store.py (hunk 4 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:197]: Done.

File: py-utils/src/utils/conf_store/conf_store.py (hunk 5 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:203]: Done. 
My understanding was wrong. I was merging both keys and values. Now only keys being merged

File: py-utils/src/utils/conf_store/conf_store.py (hunk 6 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:290]: Done

File: py-utils/src/utils/conf_store/conf_store.py (hunk 8 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:228]: Done.
My understanding was wrong. I was merging both keys and values. Now only keys being merged

File: py-utils/src/utils/conf_store/conf_store.py (hunk 9 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:229]: This code is removed

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 10]
```diff
@@ -187,6 +187,37 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        Merges the content of src_index and dest_index file
+
+        Parameters:
+        dst_index - Destination Index, to this index resulted values will be
+            merged
+        src_index - Source Index, From which we overwriting/merging key values
+        keys - optional parameter, Only specified keys will be merged.
+        """
+        if src_index not in self._cache.keys():
+            raise ConfError(errno.EINVAL, "config index %s is not loaded",
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:201]: EINVAL = Use NOT_INITIALIZED or equivalent.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 11]
```diff
@@ -187,6 +187,37 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        if keys:
+            for key in keys:
+                if self._cache[src_index].get(key):
+                    if key not in self._cache[dest_index].get_keys():
+                        self._cache[dest_index].set(key, \
+                            self._cache[src_index].get(key))
+                else:
+                    raise ConfError(errno.EINVAL, f"Invalid key {key} provided.")
+        else:
+            for key in self._cache[src_index].get_keys():
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:215]: Further the logic can be much simplified

```
def merge(self, dst, src, keys=None):
    for key in keys: 
        if not self._cache[src_index].get(key) : 
            raise ConfError(errno.ENOENT, "%s is not present in %s" %(key, src))
    if keys is None: 
        keys = self._cache[src_index].get_keys()
    self._merge(dst, src, keys)

def _merge(self, dst, src, keys): 
    for key in keys:
           if key not in self._cache[dest_index].get_keys():
               self._cache[dest_index].set(key, self._cache[src_index].get(key))
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 11 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:215]: Done

File: py-utils/src/utils/conf_store/conf_store.py [hunk 12]
```diff
@@ -187,6 +187,37 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        keys - optional parameter, Only specified keys will be merged.
+        """
+        if src_index not in self._cache.keys():
+            raise ConfError(errno.EINVAL, "config index %s is not loaded",
+                            src_index)
+        if dest_index not in self._cache.keys():
+            raise ConfError(errno.EINVAL, "config index %s is not loaded",
+                            dest_index)
+        if keys:
+            for key in keys:
```
[the remark this answers is not in the document]
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:207]: Covered in last comment

File: py-utils/src/utils/conf_store/conf_store.py (hunk 10 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:201]: Done

File: py-utils/src/utils/conf_store/conf_store.py [hunk 13]
```diff
@@ -187,6 +188,35 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        src_index - Source Index, From which we overwriting/merging key values
+        keys - optional parameter, Only specified keys will be merged.
+        """
+        if src_index not in self._cache.keys():
+            raise ConfError(errors.ERR_NOT_INITIALIZED, "config index %s is "\
+                "not loaded", src_index)
+        if dest_index not in self._cache.keys():
+            raise ConfError(errors.ERR_NOT_INITIALIZED, "config index %s is "\
+                "not loaded", dest_index)
+        if keys is None:
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:207]: This is incorrect. Sequence needs to be reversed. 
```
        for key in keys:
            if not self._cache[src_index].get(key):
                raise ConfError(errno.ENOENT, "%s is not present in %s", \
                    key, src_index)
        if keys is None:
            keys = self._cache[src_index].get_keys()
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 13 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:207]: If we loop keys in first place, we can get error if keys is None-
'TypeError: 'NoneType' object is not iterable'

Comment (ujjwalpl) [reply py-utils/src/utils/conf_store/conf_store.py:207]: You can fix as 
```
if keys is None:
    keys = self._cache[src_index].get_keys()
else: 
    for key in keys: 
        if not self._cache[src_index].get(key): 
            raise ConfError(errno.ENOENT, "key %s not present in %s", \
                key, src_index) 
```

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 13 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py:207]: Done

**Review by ujjwalpl — APPROVED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 14]
```diff
@@ -187,6 +188,36 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
             self._cache[dst_index].set(key, self._cache[src_index].get(key))
 
+    def merge(self, dest_index: str, src_index: str, keys: list = None):
+        """
+        Merges the content of src_index and dest_index file
+
+        Parameters:
+        dst_index - Destination Index, to this index resulted values will be
+            merged
+        src_index - Source Index, From which we overwriting/merging key values
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_store.py:198]: src_index - Source Index, From which new keys (and related values) are picked up for merging

File: py-utils/src/utils/conf_store/conf_store.py [hunk 15]
```diff
@@ -187,6 +188,36 @@ def copy(self, src_index: str, dst_index: str, key_list: list = None):
...
+        keys - optional parameter, Only specified keys will be merged.
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_store.py:199]: keys - optional parameter, Only these keys (and related values) from src_index will be merged.

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Please work on the comments

File: py-utils/src/utils/conf_store/conf_store.py (hunk 14 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py]: Done

File: py-utils/src/utils/conf_store/conf_store.py (hunk 15 above)
Comment (vastradparayya) [reply py-utils/src/utils/conf_store/conf_store.py]: Done

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by rahul-tripathi-git — APPROVED** [review]
Looks good to me.

**Pull request merged**